### PR TITLE
[SPARK-16463][SQL] Support `truncate` option in Overwrite mode for JDBC DataFrameWriter

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This PR adds a boolean option, `truncate`, for `SaveMode.Overwrite` of JDBC DataFrameWriter. If this option is `true`, it try to take advantage of `TRUNCATE TABLE` instead of `DROP TABLE`. This is a trivial option, but will provide great **convenience** for BI tool users based on RDBMS tables generated by Spark.

**Goal**
- Without `CREATE/DROP` privilege, we can save dataframe to database. Sometime these are not allowed for security.
- It will preserve the existing table information, so users can add and keep some additional `INDEX` and `CONSTRAINT`s for the table.
- Sometime, `TRUNCATE` is faster than the combination of `DROP/CREATE`.

**Supported DBMS**
The following is `truncate`-option support table. Due to the different behavior of `TRUNCATE TABLE` among DBMSs, it's not always safe to use `TRUNCATE TABLE`. Spark will ignore the `truncate` option for **unknown** and **some** DBMS with **default CASCADING** behavior. Newly added JDBCDialect should implement corresponding function to support `truncate` option additionally.

| Spark Dialects | `truncate` OPTION SUPPORT |
| --- | --- |
| MySQLDialect | O |
| PostgresDialect | X |
| DB2Dialect | O |
| MsSqlServerDialect | O |
| DerbyDialect | O |
| OracleDialect | O |

**Before (TABLE with INDEX case)**: SparkShell & MySQL CLI are interleaved intentionally.

``` scala
scala> val (url, prop)=("jdbc:mysql://localhost:3306/temp?useSSL=false", new java.util.Properties)
scala> prop.setProperty("user","root")
scala> df.write.mode("overwrite").jdbc(url, "table_with_index", prop)
scala> spark.range(10).write.mode("overwrite").jdbc(url, "table_with_index", prop)
mysql> DESC table_with_index;
+-------+------------+------+-----+---------+-------+
| Field | Type       | Null | Key | Default | Extra |
+-------+------------+------+-----+---------+-------+
| id    | bigint(20) | NO   |     | NULL    |       |
+-------+------------+------+-----+---------+-------+
mysql> CREATE UNIQUE INDEX idx_id ON table_with_index(id);
mysql> DESC table_with_index;
+-------+------------+------+-----+---------+-------+
| Field | Type       | Null | Key | Default | Extra |
+-------+------------+------+-----+---------+-------+
| id    | bigint(20) | NO   | PRI | NULL    |       |
+-------+------------+------+-----+---------+-------+
scala> spark.range(10).write.mode("overwrite").jdbc(url, "table_with_index", prop)
mysql> DESC table_with_index;
+-------+------------+------+-----+---------+-------+
| Field | Type       | Null | Key | Default | Extra |
+-------+------------+------+-----+---------+-------+
| id    | bigint(20) | NO   |     | NULL    |       |
+-------+------------+------+-----+---------+-------+
```

**After (TABLE with INDEX case)**

``` scala
scala> spark.range(10).write.mode("overwrite").option("truncate", true).jdbc(url, "table_with_index", prop)
mysql> DESC table_with_index;
+-------+------------+------+-----+---------+-------+
| Field | Type       | Null | Key | Default | Extra |
+-------+------------+------+-----+---------+-------+
| id    | bigint(20) | NO   | PRI | NULL    |       |
+-------+------------+------+-----+---------+-------+
```

**Error Handling**
- In case of exceptions, Spark will not retry. Users should turn off the `truncate` option.
- In case of schema change:
  - If one of the column names changes, this will raise exceptions intuitively.
  - If there exists only type difference, this will work like Append mode.
## How was this patch tested?

Pass the Jenkins tests with a updated testcase.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Due to the difference of scope, I create another JIRA issue.

**Comment by dongjoon-hyun** [comment]
Could you review this PR, @srowen ?

**Comment by srowen** [comment]
Hm, it seems like it should be another `SaveMode` if anything. Its semantics would be identical to `Overwrite`, I guess, for other non-JDBC sources. CC @yhuai ?

However... is it not possible to just TRUNCATE when the schema hasn't changed, and DROP/CREATE when it has? that seems like the best solution if it's feasible.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @srowen . Currently, the `truncate` option will be ignored for other non-JDBC sources.
IMO, most of them is **file-based* sources, so they do not need this.

Indeed, DROP/CREATE is the best and robust solution, but, as you see the example, this is a minor optimization for **good** benefit. We can not achieve the benefit with `DROP`.

**Comment by srowen** [comment]
Shouldn't truncate function like overwrite in instances where the distinction doesn't matter -- rather than be ignored?

I think the point of the issue was that DROP/CREATE _isn't_ the best solution in all cases, specifically, where the table schema has not changed. It's at best a little slower and at worst destroys metadata.

I believe everyone sees there are use cases for both TRUNCATE and DROP/CREATE; I'm more specifically asking if we need to make the caller figure this out or whether it's easy and simple to use TRUNCATE when possible in Overwrite mode. Maybe it's not simple, or maybe we want to let people control this anyway, but it does have this cost of adding another setting to the API, and it's still possible to surprise yourself with Overwrite mode in this context if you're not aware of what Truncate does differently.

**Comment by dongjoon-hyun** [comment]
That's a good point of view. Right. I agree with you, `DROP/CREATE isn't the best solution in all cases`. We can automatically check the schema **compatibility** in some way. 
Let me dig more on this issue. It seems not simple, but looks more valuable.

**Comment by dongjoon-hyun** [comment]
Thank you for guiding me. I hope this PR become more valuable in Spark, too.

**Comment by dongjoon-hyun** [comment]
Hi, @srowen .
I'm looking at your advice more. Here is my summary so far in general.

**SaveMode.TRUNCATE**
- Truncate the content only, not a structure of existing data source.
- I prefer this mode raises exception if there is a problem.
  (I mean it's not safe for the users to have `SaveMode.Overwrite` as a the fallback operation.)
- The (SCC)  is the key. 

**Behavior (how to handle an check schema compatibility)**
- JDBC: USE `TRUCATE TABLE` if the source has the same dialect (SAME DB), the same schema(name, type, nullability, order).

- File-based Source(like JSON): DELETE FILE ONLY if the inferred schema is identical  (Directory means partitions.)

Dose this make sense to you? If I missed some of your advice, please let me know.

**Comment by srowen** [comment]
I suppose my key question is still: do we need to make the user choose at all? it seems like TRUNCATE is always the right choice except when schema changes. That means, ideally, no new SaveMode. Overwrite would just cause TRUNCATE where possible, otherwise DROP/CREATE.

I think this is specific to JDBC only. Other sources can leave their behavior unchanged.

Yes, I think the next question is, how hard is it to reliably detect that the schema hasn't changed? you know better than I, likely.

**Comment by dongjoon-hyun** [comment]
`TRUNCATE` is not always right choice for the existing tables, because sometime users really wants to delete all information including INDEX/PARTITIONS, too.
`SaveMode.Overwrite` should mean `SaveMode.Overwrite`. We can not optimize it without user permission even for the same schema .

**Comment by srowen** [comment]
It's possible. I think it's probably much more likely the metadata is supposed to be kept, but who knows. OK this has a decent logic. I'd prefer to get someone else's opinion who has worked on this code like .. @lw-lin or @zsxwing ?

**Comment by dongjoon-hyun** [comment]
Just rebased to the master in order to ensure it works.

**Comment by srowen** [comment]
This seems pretty reasonable. Anyone else have an opinion on offering this option?

**Comment by dongjoon-hyun** [comment]
Thank you for keeping attention this PR, @srowen .

**Comment by gatorsmile** [comment]
Let me share my 2 cents here.

`Truncate Table` is very risky. It is fast since RDBMS does not log the individual row deletes. That means, we are unable to roll it back in most RDBMS. The behaviors are different in different RDBMS vendors. For example, 

1. [PostgreSQL](https://www.postgresql.org/docs/9.1/static/sql-truncate.html) could truncate all its descendant tables at default, unless we specify `ONLY` in the statement.

2. [DB2 z](https://www.ibm.com/support/knowledgecenter/SSEPEK_10.0.0/sqlref/src/tpc/db2z_sql_truncate.html) can roll it back but the table is still in a truncated state

3. Oracle 11 documents this caution at the very beginning: http://docs.oracle.com/cd/B28359_01/server.111/b28286/statements_10007.htm Unable to roll it back. (Note, [Oracle 12 ](ttps://docs.oracle.com/database/121/SQLRF/statements_10007.htm#SQLRF01707) added a new clause CASCADE. By default, the descendant table will not be dropped.)

Thus, I am not confident to add such a mode or option into Spark SQL. This command is used very rarely in production systems. DBAs should do it manually, instead of using the Spark SQL interface.

If the community still thinks we should integrate it into Spark SQL, I am also fine. After a quick review, I think the implementation misses the error handling. For example, when users try to use `truncate` option in the other `SaveMode`.

**Comment by srowen** [comment]
@gatorsmile that all sounds reasonable, but right now a DROP/CREATE table happens. That's also not possible within a transaction and is a more drastic operation. Does this argument not apply more to existing behavior? the point indeed is to perform a more modest operation if possible.

**Comment by dongjoon-hyun** [comment]
Thank you for attention, @gatorsmile .

BTW, this option is for the advanced users who knows their DB and the limitation and powers of `TRUNCATE`.

The the following comment, I really like this point of views. And I'm sure that DBAs will allow the use of this option to Spark Architect/Developer/Users for the case they can do manually.
> DBAs should do it manually, instead of using the Spark SQL interface.

**Comment by dongjoon-hyun** [comment]
Yep. I totally agree with @srowen 's opinions, too.

**Comment by dongjoon-hyun** [comment]
Just rebased to the master.

**Comment by dongjoon-hyun** [comment]
According to the code, this PR changed the code of @rxin .

Hi, @rxin . 
Could you give us some opinion about supporting `truncate` option with SaveMode.Overwrite for JDBC sources?

**Comment by gatorsmile** [comment]
- `DROP`/`CREATE`/`DELETE` can be rolled back. In Oracle, we are unable to roll back `TRUNCATE`. I am not sure whether this is a big deal to the Spark users.

- If the data source is `PostgreSQL`, we also truncate all its descendant tables by default. This side effect does not look good to me. Not sure whether the Spark users can tolerate it.

- So far, Spark does not have role-based privilege. It completely depends on the underlying data sources. If DBAs have to grant higher privileges to Spark app (for any reason), this change could be a potential issue. 

I did not investigate all the RDBMS vendors and different versions might have different behaviors. My suggestion is to do more investigation before adding this support.

**Comment by dongjoon-hyun** [comment]
The followings are my opinions according to the priority.
* First of all, `TRUNCATE` is the lower privilege than DROP/CREATE.
  - You know that DROP/CREATE privilege do anything harm.
* Second, we are assuming the original table is generated by `DataFrameWriter`, not any other general database table.
  - Why do you consider the descendant tables? If you consider `descendant tables`, you should say `DROP` will also delete them all, too.
* Third, do we use rollback with `DataFrameWriter`?

I'm worrying about losing the focus. The all the concerns are based on the correct facts, but the scope of arguments seems to be slightly too general to this PR. Please see the description of this PR. The context of this PR is **providing truncate option for the Spark JDBC tables generated by df.write.mode("overwrite").jdbc(url, "table_with_index", prop)**.

**Comment by gatorsmile** [comment]
> `TRUNCATE` is the lower privilege than DROP/CREATE. 
This is not true. For example, in DB2 z, `DROP` and `TRUNCATE` require different privilege.
https://www.ibm.com/support/knowledgecenter/SSEPEK_10.0.0/sqlref/src/tpc/db2z_sql_truncate.html 
https://www.ibm.com/support/knowledgecenter/SSEPEK_10.0.0/sqlref/src/tpc/db2z_sql_drop.html

- Drop table will not drop the descendant tables, unless you explicitly specify `CASCADE`. The behaviors of `TRUNCATE` depends on the vendors. `TRUNCATE` is not included in the SQL ANSI standards until 2008. That might be the reason why different vendors behave very differently. 

- Rollback may be not an issue here, but it means the deletion log records are not written into the log. Does it matter? It really depends on how the applications use the output table of `DataFrameWriter` and the recovery log they generated. The product I previously worked is to parse and utilize the transaction recovery log of different RDBMS. Thus, I knew `Truncate` is a big issue for most replication products.

I just share what I learned here. Developing a general solution for different JDBC data sources is always very very complicated. When more and more enterprise customers start using Spark, we will get more and  more `strange` JIRAs.

**Comment by gatorsmile** [comment]
I think the above discussions are important when users really use the option `truncate`. I am OK to let users decide whether `TRUNCATE` is good or not for their scenario.

**Comment by dongjoon-hyun** [comment]
`Lower` does not mean `inclusion`. Those are different privileges of course. But, I agree with you here.
> I just share what I learned here. Developing a general solution for different JDBC data sources is always very very complicated. When more and more enterprise customers start using Spark, we will get more and more strange JIRAs.

Let me improve more this PR. We have JDBC Dialect for that reason.

**Comment by dongjoon-hyun** [comment]
I always appreciate your intensive reviews. Those always make my PRs meaningful and stronger. Thank you, @srowen and @gatorsmile .

**Comment by dongjoon-hyun** [comment]
To identify further works, I'll rephrase your comments.

You means the case `SaveMode.Overwrite` should fail due to that current Spark uses `statement.executeUpdate(s"DROP TABLE $table")`.

 > Drop table will not drop the descendant tables, unless you explicitly specify CASCADE. The behaviors of TRUNCATE depends on the vendors.

In this case, what you want in this PR is `truncate` option also should fail in the same way. Right?

**Comment by dongjoon-hyun** [comment]
By the way, I'm just wondering. Does IBM STC has some test environments for various DBMS? I guess you should have that. :)

**Comment by gatorsmile** [comment]
Nope... I lost the IDs to access various DBMS after joining STC. Previously, I had. : )

**Comment by dongjoon-hyun** [comment]
Thank you for solving my curiosity! I'm collecting the behavior of TRUNCATE default to ensure robustness.

**Comment by gatorsmile** [comment]
You can try to use Dockers. You can find most RDBMS, including Hive, Oracle and DB2 on LUW. The only exception is DB2 on z/OS, which is running on mainframe.

**Comment by dongjoon-hyun** [comment]
Thanks. I'll try to build some!

**Comment by srowen** [comment]
Yeah interesting discussion for sure. It sounds like it's not valid to assume TRUNCATE is OK if DROP/CREATE is OK. It sounds like it's useful to maybe let the user choose via a little config option here.

**Comment by dongjoon-hyun** [comment]
I think it's valuable to explicitly compare `truncate` option with the existing `SaveMode.Overwrite` in popular DBMSs.
After some testing, we can provide this with more clear limitation notes if exists.

**Comment by dongjoon-hyun** [comment]
Now, the PR have a ability to allow `truncate` option only if `TRUNCATE TABLE` default is not cascading. For the unknown or known-to-be-cascading-default DBMS (Postgres), `truncate` option is ignored.

Spark Dialects | CASCADING TRUNCATE DEFAULT | `truncate` OPTION SUPPORT
---------------|-----------------------------------|-------------------------------
MySQLDialect | X | O
PostgresDialect | O | X
DB2Dialect | X | O
MsSqlServerDialect | X | O
DerbyDialect | X | O
OracleDialect | X | O
H2Dialect | X | O

During searching, I found this [DBMS Comparison](http://www.sql-workbench.net/dbms_comparison.html) (Last updated: 2016-05-16)

It looks consistent with each DBMS's docs. Please note that the approach of this PR gives two steps of choice for users.
- First step, users can choose `truncate` option for their own risk.
- Second step, although the `truncate` option is true, we do not use `TRUNCATE TABLE` for unknown DBMSs or known-to-be-different DBMS.
Always, the fallback operation is the existing `DROP TABLE`.

**Comment by dongjoon-hyun** [comment]
Ur, it's strange. `JDBCSuite` is passed locally. I'll rebase this.

**Comment by dongjoon-hyun** [comment]
Yep. The old failure seems due to Jenkins flakiness.
It's ready for review again. Could you review this PR when you have sometime, @srowen and @gatorsmile ?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 1]
```diff
@@ -419,8 +422,13 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       }
 
       if (mode == SaveMode.Overwrite && tableExists) {
-        JdbcUtils.dropTable(conn, table)
-        tableExists = false
+        if (extraOptions.getOrElse("truncate", "false").toBoolean &&
+            JdbcUtils.isCascadingTruncateTable(url) == Some(false)) {
+          JdbcUtils.truncateTable(conn, table)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: If `truncateTable` failed due to a non fatal exception, should we fall back to the previous way (i.e., drop and create)? This is a design decision. CC @srowen @rxin

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I'd say no, because user has explicitly specified truncate. They can turn if off themselves.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: : ) Sure. Then, the next design question to @rxin and @srowen 

Should we still truncate the table if the table schema does not match the schema of new table?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: We should do whatever we do with drop.

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: First of all, it will raise exceptions if one of the column names are different.
For the different column type with same column name, it will work like `SaveMode.Append` operation.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I see. Then, the current implementation looks good to me. 

@dongjoon-hyun Could you summarize the previous discussion and design decision we made? Document them in the PR description. Thanks!

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Currently, it raises exceptions if one of the column names is different.
For the different column type with same column name, it works like `SaveMode.Append` operation.
I thought for the trade-off between DROP and TRUNCATE.

Let me think about the decision point.

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: For my understanding, I will ask one question. 

Literally, we should not do whatever we do with drop, e.g., we should not drop INDEX, right?

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Sure. I'll update the document and PR description more clearly.

Thank you for guidance, @rxin and @gatorsmile .

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: - Drop, Create and Insert: Create and Insert could fail, but we still drop the table. 
- Truncate and Insert: Insert could fail, but we always truncate the table. 

I think it is OK to raise an exception here, but check whether the exception message is meaningful or not.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Nope, dropping index does not make sense here.

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: The current exception message is "Column `seq` not found".

**Comment by dongjoon-hyun** [comment]
The descriptions of PR/code are updated.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 2]
```diff
@@ -390,6 +390,12 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
    * You can set the following JDBC-specific option(s) for storing JDBC:
    * <li>`truncate` (default `false`): use `TRUNCATE TABLE` instead of `DROP TABLE`.</li>
    *
+   * In case of failures, users should turn off `truncate` option to use `DROP TABLE` again. Also,
+   * due to the different behavior of `TRUNCATE TABLE` among DBMS, it's not always safe to use this.
+   * MySQLDialect, DB2Dialect, MsSqlServerDialect, DerbyDialect, DerbyDialect supports this option
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Please correct this.

**Comment by gatorsmile** [comment]
LGTM except one minor comment.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 2 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Oh, sure.

**Comment by dongjoon-hyun** [comment]
Thank you, @gatorsmile .
I removed the repeated `DerbyDialect` and adds `OracleDialect`.

**Comment by gatorsmile** [comment]
LGTM

**Comment by dongjoon-hyun** [comment]
Hi, @srowen .
Could you review this JDBC PR again?

**Comment by dongjoon-hyun** [comment]
Rebased.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCWriteSuite.scala [hunk 3]
```diff
@@ -145,14 +153,24 @@ class JDBCWriteSuite extends SharedSQLContext with BeforeAndAfter {
...
+    val df3 = spark.createDataFrame(sparkContext.parallelize(arr2x3), schema3)
 
     df.write.jdbc(url1, "TEST.TRUNCATETEST", properties)
-    df2.write.mode(SaveMode.Overwrite).jdbc(url1, "TEST.TRUNCATETEST", properties)
+    df2.write.mode(SaveMode.Overwrite).option("truncate", true)
+      .jdbc(url1, "TEST.TRUNCATETEST", properties)
     assert(1 === spark.read.jdbc(url1, "TEST.TRUNCATETEST", properties).count())
     assert(2 === spark.read.jdbc(url1, "TEST.TRUNCATETEST", properties).collect()(0).length)
+
+    val m = intercept[SparkException] {
```
Comment (srowen) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCWriteSuite.scala]: To check my understanding here, this overwrites the table with a different schema (new column `seq`). This shows the truncate fails because the schema has changed.

I guess it would be nice to test the case where the truncate works at least, though, we can't really test whether it truncates vs drops.

Could you for example just repeat the code on line 163-166 here to verify that overwriting just results in the same results?

Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCWriteSuite.scala]: Sure, that would be better.

Comment (srowen) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCWriteSuite.scala]: Ah, nevermind my last comment. You test the truncate succeeds path already.
OK, the assertion here makes sense though it highlights that if truncation can't succeed, then it only fails after truncating and the new dataframe can't be written. I suppose that's reasonable semantics, since it otherwise requires doing something like testing an insert.

**Comment by srowen** [comment]
Merged to master

**Comment by dongjoon-hyun** [comment]
Thank you for review and merging, @srowen , @rxin , and @gatorsmile !
